### PR TITLE
Add track move commands to controls library

### PR DESCRIPTION
**Pull request by nlw0**
Add track move to controls library by emulating alt+up/down. Depends on #13092

### CODE REVIEW

File: src/library/librarycontrol.cpp [hunk 1]
```diff
@@ -793,6 +816,25 @@ void LibraryControl::slotMoveFocus(double v) {
...
+    if (v > 0) {
+        slotMoveTrack(-1);
+    }
+}
+
+void LibraryControl::slotMoveTrack(double v) {
+    const auto key = (v < 0) ? Qt::Key_Up : Qt::Key_Down;
+    const auto times = static_cast<unsigned short>(std::abs(v));
+    emitKeyEvent(QKeyEvent{
+            QEvent::KeyPress, key, Qt::AltModifier, QString(), false, times});
```
Comment (ronso0) [inline src/library/librarycontrol.cpp:835]: unfortunately emitKeyEvent requires the library to be focused, which would limit the use cases (same with the current navigation controls, see #9294, #11285 and others).

Let's send the key event to `m_pLibraryWidget` directly as this is the only widget that would accept this event (check if is not null)

**Review by ronso0 — CHANGES_REQUESTED** [review]
Thanks for this PR!

Please sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

Code LGTM, just a comment

File: src/library/librarycontrol.cpp (hunk 1 above)
Comment (nlw0) [reply src/library/librarycontrol.cpp]: All right, I couldn't test yet, does the code look right?

**Comment by nlw0** [comment]
Sure. I believe I signed it already, though.

File: src/library/librarycontrol.cpp [hunk 2]
```diff
@@ -166,6 +166,29 @@ LibraryControl::LibraryControl(Library* pLibrary)
                 &LibraryControl::slotMoveFocus);
     }
 
+    // Controls to move tracks on playlist
+    // Track move controls (emulate Alt+up/down button press)
+    m_pMoveTrackForward = std::make_unique<ControlPushButton>(ConfigKey("[Library]", "MoveTrackForward"));
+    m_pMoveTrackBackward = std::make_unique<ControlPushButton>(ConfigKey("[Library]", "MoveTrackBackward"));
+    m_pMoveTrack = std::make_unique<ControlEncoder>(ConfigKey("[Library]", "MoveTrack"), false);
```
Comment (ronso0) [inline src/library/librarycontrol.cpp:173]: I suggest we make it
MoveTracks | ~Up | ~Down

just FYI: we aim to use snake_case for new controls.
Here PascalCase is okay as it it's consistent with the closely related controls.

File: src/library/librarycontrol.cpp [hunk 3]
```diff
@@ -793,6 +816,31 @@ void LibraryControl::slotMoveFocus(double v) {
...
+    }
+}
+
+void LibraryControl::slotMoveTrackBackward(double v) {
+    if (v > 0) {
+        slotMoveTrack(-1);
+    }
+}
+
+void LibraryControl::slotMoveTrack(double v) {
```
Comment (ronso0) [inline src/library/librarycontrol.cpp:831]: Please add a  /// comment here explaining that these moves equal Alt + Up/Down.

**Review by ronso0 — CHANGES_REQUESTED** [review]
LGTM, just the names need to be changed.

File: src/library/librarycontrol.cpp (hunk 2 above)
Comment (nlw0) [reply src/library/librarycontrol.cpp:173]: done

File: src/library/librarycontrol.cpp (hunk 3 above)
Comment (nlw0) [reply src/library/librarycontrol.cpp:831]: done

File: src/library/librarycontrol.cpp (hunk 2 above)
Comment (ronso0) [reply src/library/librarycontrol.cpp]: > MoveTracks | ~Up | ~Down

Ha, sorry that was not clear, I meant:
MoveTrack**s**
MoveTrack**s**Up
MoveTrack**s**Down

**Comment by ronso0** [comment]
pre-commit is complaining (didn't take a closer look)

**Comment by ronso0** [comment]
Sorry, my mistake:
`m_pLibraryWidget` is the stacked widget that holds all WTrackTableViews, it won't handle the keypress.

I'll prepare a commit that allows to send the keypress to the current view (if it's a WTrackTableView).

File: src/library/librarycontrol.cpp [hunk 4]
```diff
@@ -166,6 +166,30 @@ LibraryControl::LibraryControl(Library* pLibrary)
                 &LibraryControl::slotMoveFocus);
     }
 
+    // Controls to move tracks on playlist
+    // Track move controls (emulate Alt+up/down button press)
+    m_pMoveTrackUp = std::make_unique<ControlPushButton>(ConfigKey("[Library]", "MoveTrack-Up"));
+    m_pMoveTrackDown = std::make_unique<ControlPushButton>(
+            ConfigKey("[Library]", "MoveTrack-Down"));
```
Comment (ronso0) [inline src/library/librarycontrol.cpp:173]: Let's rename so the controls match the scheme of `MoveUp`, `ScrollUp` etc.
```suggestion
    m_pMoveTrackUp = std::make_unique<ControlPushButton>(ConfigKey("[Library]", "MoveTrackUp"));
    m_pMoveTrackDown = std::make_unique<ControlPushButton>(
            ConfigKey("[Library]", "MoveTrackDown"));
```

**Review by ronso0 — CHANGES_REQUESTED** [review]

**Comment by ronso0** [comment]
If you adopt my code suggestion, then rebase (and squash commits) onto main then there is `WLibrary::getCurrentTrackTableView()` available, and you can pick (and also squash) this commit
ab8c1d21ae
Works nicely!

**Comment by ronso0** [comment]
@nlw0 Do you have time to rebase, cherry-pick and squash as I suggested above?

Let us know if you don't, then I'd take over soonish in order to get this into 2.5 along with the keyboard improvement.

**Comment by nlw0** [comment]
Hi, I can look into it again, but I wasn't able to use that commit, how do I check it out? On GitHub it says `This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.`

**Comment by nlw0** [comment]
The SlotMoveTrack commands don't seem to be on the 2.5 branch yet, should I wait for another branch to be merged firsts?

**Comment by ronso0** [comment]
yes, the Alt moves are available in 2.5 https://github.com/mixxxdj/mixxx/blob/2.5/src/widget/wtracktableview.cpp#L867

I think you need to `git fetch [mixxx upstream]`, you'll see that 2.5 has been updated since your last commit.
Just squash your commits so we have only one essential commit with extra the CI fixup commits, then rebase. Since this is a PR commit that should work flawlessly.

**Comment by nlw0** [comment]
All right I haven''t tested yet but here it is. Hope I didn't mess up anything.

**Comment by ronso0** [comment]
Yep, looks good, thank you!

**Comment by ronso0** [comment]
Could you also take care of documenting the new controls in the manual? : )
https://manual.mixxx.org/2.5/en/chapters/appendix/mixxx_controls.html#the-library-controls

**Pull request merged**